src/coreclr/jit/fgdiagnostic.cpp

```diff
@@ -3563,6 +3564,474 @@ void Compiler::fgDebugCheckNodesUniqueness()
+        {
+            // All uses of ssa vars should have valid ssa nums, unless there are no defs.
+            //
+            if (!tree->HasSsaName())
+            {
+                if (varDsc->lvPerSsaData.GetCount() > 0)
+                {
+                    SetHasErrors();
+                    JITDUMP("[error] Missing SSA number on use [%06u] (V%02u)\n", m_compiler->dspTreeID(tree), lclNum);
+                }
+                return fgWalkResult::WALK_CONTINUE;
+            }
+
+            const unsigned ssaNum = tree->AsLclVarCommon()->GetSsaNum();
+            SsaKey         key(lclNum, ssaNum);
+            SsaInfo*       ssaInfo = nullptr;
+
+            if (!m_infoMap.Lookup(key, &ssaInfo))
+            {
+                ssaInfo = new (m_compiler->getAllocator(CMK_DebugOnly)) SsaInfo;
+                m_infoMap.Set(key, ssaInfo);
+            }
+
+            if (tree->OperIs(GT_PHI_ARG))
+            {
+                ssaInfo->AddPhiUse(m_block, key);
+            }
+            else
+            {
+                ssaInfo->AddUse(m_block, key);
+            }
+        }
+
+        return fgWalkResult::WALK_CONTINUE;
+    }
+
+    void SetHasErrors()
+    {
+        if (!m_hasErrors)
+        {
+            JITDUMP("fgDebugCheckSsa: errors found\n");
+            m_hasErrors = true;
+        }
+    }
+
+    bool HasErrors() const
+    {
+        return m_hasErrors;
+    }
+
+    void SetBlock(BasicBlock* block)
+    {
+        m_block = block;
+    }
+
+    void DoChecks()
+    {
+        for (unsigned lclNum = 0; lclNum < m_compiler->lvaCount; lclNum++)
+        {
+            LclVarDsc* const varDsc = m_compiler->lvaGetDesc(lclNum);
+
+            if (!varDsc->lvInSsa)
+            {
+                continue;
+            }
+
+            const SsaDefArray<LclSsaVarDsc>& ssaDefs = varDsc->lvPerSsaData;
+
+            for (unsigned i = 0; i < ssaDefs.GetCount(); i++)
+            {
+                LclSsaVarDsc* const ssaVarDsc = ssaDefs.GetSsaDefByIndex(i);
+                const unsigned      ssaNum    = ssaDefs.GetSsaNum(ssaVarDsc);
+
+                SsaKey   key(lclNum, ssaNum);
+                SsaInfo* ssaInfo = nullptr;
+
+                if (!m_infoMap.Lookup(key, &ssaInfo))
+                {
+                    // Possibly there are no more references to this local.
+                    continue;
+                }
+
+                // VarDsc block should be accurate.
+                //
+                BasicBlock* const ssaInfoDefBlock   = ssaInfo->GetDefBlock();
+                BasicBlock* const ssaVarDscDefBlock = ssaVarDsc->GetBlock();
+
+                if (ssaInfoDefBlock != ssaVarDscDefBlock)
+                {
+                    // We are inconsistent in tracking where the initial values of params
+                    // and uninit locals come from. Tolerate.
+                    //
+                    const bool initialValOfParamOrLocal = (lclNum < m_compiler->lvaCount) && (ssaNum == 1);
```

Comment:
> `(lclNum < m_compiler->lvaCount)`

Isn't this always `true`?

Reply:
Thanks for spotting this. It was meant to track the cases where SSA creates a virtual def, but I don't yet have the right checks for that.

Reply:
FWIW, the current exclusions look right to me.

Notably, there exists ambiguity between a virtual definition, and a real definition made by a call in the first block (because calls don't get recorded in `m_asg`, even as they perhaps should), but that does not seem relevant here.

Reply:
Yeah, that whole area is a bit messy. Assuming no other feedback I may leave this as is and clean it up later.